espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/espresso/EspressoW3CActionAdapter.java

```diff
@@ -119,6 +168,158 @@ public void sychronousTickActionsComplete() throws AppiumException {
         multiTouchState.perform(uiController);
     }
 
+    public int getKeyCode(String keyValue, int location) throws AppiumException {
```

Comment:
This implementation is ok, but I the supported chars range is pretty limited. We could use the same approach as adb though: https://android.googlesource.com/platform/frameworks/base.git/+/8f5e3d13e166e4873e130a5ef7c987333dbac917/cmds/input/src/com/android/commands/input/Input.java#148

Reply:
Nice. Does that approach provide a full mapping between unicode and the key events, because ideally I would like to not have the number values.

Reply:
I assume it provides the mapping between the currently selected IME. You could try it on your own by just executing `adb shell input text` with different values

Reply:
I just tried the ADB way and it seems to work fine. This is a much better way.